Handle exif face regions with zero embeddings

kvalev opened the pull request:
When facenet cannot compute an embeddings vector for a given exif face region, the current implementation will simply skip the region, as PhotoPrism simply cannot work with faces without embeddings.

This PR implements a fallback for such cases - run face detection for the given image, including low quality detections and check whether any of the detections significantly overlaps with the exif region. Face detection is done on the original image and not on a thumbnail in order to have as many detections as possible.

Additionally the user can configure a list of angles at which aforementioned fallback face detection should be ran. Per default, the image will not be rotated, but if the user has a lot of exif face regions at non-standard angles, a list of angles can be configured using the `PHOTOPRISM_FACE_REGION_ANGLES` environment variable. Example:

```
PHOTOPRISM_FACE_REGION_ANGLES=0,90,180,270
```

related to #59

kvalev committed:
People: Fallback scenario for exif face regions with zero embeddings

When facenet cannot compute an embeddings vector for a given exif face region,
the current implementation will simply skip the region, as PhotoPrism simply
cannot work with faces without embeddings.
This commit implements a fallback for such cases - run face detection for the
given image, including low quality detections and check whether any of the
detections significantly overlaps with the exif region.

kvalev commented:
@rickysarraf This PR should be now in a good shape and you can give it a go if you want. I evaluated it with one of your test pictures that you sent me, where the face region was very small and photoprism could not find any embeddings and with this PR the face region is found and indexed correctly. The key was do repeat the face detection with the original image and including "low quality" faces, which found approximately the same face region (with slightly face area).

rickysarraf commented:
> @rickysarraf This PR should be now in a good shape and you can give it a go if you want. I evaluated it with one of your test pictures that you sent me, where the face region was very small and photoprism could not find any embeddings and with this PR the face region is found and indexed correctly. The key was do repeat the face detection with the original image and including "low quality" faces, which found approximately the same face region (with slightly face area).

@kvalev 

I now have some test results for a set of images accumulated from 6 months ago, all over 2023. I will now also run a new test on some of the older, scanned images. Those may provide with more interesting results. I will share the link privately over email.

kvalev commented:
I evaluated the logs and the numbers are as follows:

45 face regions had zero embeddings
6 out of those could be matched to a similar region produced by the face detection
3 out of those had a embedding

Meaning only 3/45 face could be recovered with the implemented fallback, which is not good. Based on the logs I selected a couple of files, which will be interesting for debugging. Would it be possible to send them, so that I can see what (if anything) can be improved?

20230103_104543_D6005F9C.jpg
20230306_171014_61F90C87.jpg
20230323_040840_21967141.jpg
20230323_040845_39B5372C.jpg
20230323_040854_6FEED0F7.jpg
20230222_145245_76CE1DAF.jpg

rickysarraf commented:
> I evaluated the logs and the numbers are as follows:
> 
> 45 face regions had zero embeddings 6 out of those could be matched to a similar region produced by the face detection 3 out of those had a embedding
> 
> Meaning only 3/45 face could be recovered with the implemented fallback, which is not good. Based on the logs I selected a couple of files, which will be interesting for debugging. Would it be possible to send them, so that I can see what (if anything) can be improved?
> 
> 20230103_104543_D6005F9C.jpg 20230306_171014_61F90C87.jpg 20230323_040840_21967141.jpg 20230323_040845_39B5372C.jpg 20230323_040854_6FEED0F7.jpg 20230222_145245_76CE1DAF.jpg

THank you @kvalev 

I have emailed you with a link to the photos.

kvalev committed:
People: Remove support for comparing faces of different scales

In the first implementation draft, we had face regions of different scales -
some were created from the original image and some - from the thumbnail. As
such, the regions had different scales, which had to be taken into account when
comparing them.
Now all of the regions are of the same scale - that of the original image, so
the normalization process is superfluous.

kvalev committed:
People: Fallback scenario for angled face regions

The most probable reason why a manually labelled face region could not be
detected by PhotoPrism seems to be that the face region is at a weird angle. In
order to support such face regions, a fallback has been added, which kicks in
when no embeddings can be computed for a manually labelled face region. In this
case, the fallback runs the face detection at a couple of hardcoded angles
(close to 90°, 180° and 270°) and checks whether a face has been detected, which
significatly overlaps with the manual face region.

kvalev commented:
@rickysarraf Can you try out the latest version from this branch when you get the chance? Almost all face regions should be recognized now - from the sample files that you sent me only one is not recognized, but I have a fix for this one as well.

The reason why previously so many face regions were not detected properly was that the faces are at an angle (judging by the sample files). For example the faces in `20230323_040845_39B5372C.jpg` are at an -90°/270° angle. PhotoPrism per default is quite bad a detecting such "rotated" faces, as the face detection is configured to only run at 0°. My workaround was to run it additionally for few hardcoded angles (approx 90° and 270°) to try and detect such "rotated" faces. This is only done for the exif face regions and only in case PhotoPrism does not detect a face or cannot compute an embedding for the given region. As such, this additional overhead should be reduced to a minimum, but it still might be significant, depending on how many such face regions there are and how many different angles I end up hardcoding.

rickysarraf commented:
> @rickysarraf Can you try out the latest version from this branch when you get the chance? Almost all face regions should be recognized now - from the sample files that you sent me only one is not recognized, but I have a fix for this one as well.
> 
> The reason why previously so many face regions were not detected properly was that the faces are at an angle (judging by the sample files). For example the faces in `20230323_040845_39B5372C.jpg` are at an -90°/270° angle. PhotoPrism per default is quite bad a detecting such "rotated" faces, as the face detection is configured to only run at 0°. My workaround was to run it additionally for few hardcoded angles (approx 90° and 270°) to try and detect such "rotated" faces. This is only done for the exif face regions and only in case PhotoPrism does not detect a face or cannot compute an embedding for the given region. As such, this additional overhead should be reduced to a minimum, but it still might be significant, depending on how many such face regions there are and how many different angles I end up hardcoding.

@kvalev I have sent you an email with link to the logs.  I felt the overhead was okay, given that it didn't run the angular detection logic on every file.

rickysarraf commented:
@kvalev I'm also persistently seeing the index error on my builds.

```
photoprism_1  | 2023-07-07T08:40:15.062528154Z time="2023-07-07T08:40:15Z" level=warning msg="index: invalid ID specified (update covers)"
```


I tested this on your build and did not see the problem.

```
DEBU[2023-07-07T08:54:38Z] purge: updated index [20.312651796s]         
INFO[2023-07-07T08:54:38Z] purged 104 files and 101 photos in 3m35.461097942s 
DEBU[2023-07-07T08:54:38Z] closed database connection                   
```

So the only extra PRs in my build are: #135 #138 #127

kvalev commented:
@rickysarraf Do you have any additional logs from the indexing operation? I assume its related to #127, as its not quite done yet.

rickysarraf commented:
> @rickysarraf Do you have any additional logs from the indexing operation? I assume its related to #127, as its not quite done yet.

Hello @kvalev 

Thanks for the clarification. I was indeed using #127  as well.



Here's logs with some extra information.


```
photoprism_1  | 2023-07-09T17:54:40.477886416Z time="2023-07-09T17:54:40Z" level=debug msg="photos: found 23,036 results for filter:\"public:true country:in\" public:true dist:20 country:in [23.603056226s]"
photoprism_1  | 2023-07-09T17:54:40.480294282Z time="2023-07-09T17:54:40Z" level=info msg="moments: state album 'India 1977' is below threshold, but will not be deleted"
photoprism_1  | 2023-07-09T17:54:40.726402026Z time="2023-07-09T17:54:40Z" level=info msg="moments: label album 'Bays, Capes & Beaches' is not empty, has 57 photos"
photoprism_1  | 2023-07-09T17:54:40.848270010Z time="2023-07-09T17:54:40Z" level=info msg="moments: label album 'Outdoor Adventures' is not empty, has 84 photos"
photoprism_1  | 2023-07-09T17:54:40.969967315Z time="2023-07-09T17:54:40Z" level=info msg="moments: label album 'Water, Ships & Boats' is not empty, has 44 photos"
photoprism_1  | 2023-07-09T17:54:41.091690930Z time="2023-07-09T17:54:41Z" level=info msg="moments: label album Pets is not empty, has 23 photos"
photoprism_1  | 2023-07-09T17:54:41.235604520Z time="2023-07-09T17:54:41Z" level=info msg="moments: label album 'Festivals & Entertainment' is not empty, has 18 photos"
photoprism_1  | 2023-07-09T17:54:41.402154505Z time="2023-07-09T17:54:41Z" level=info msg="moments: label album 'Nature & Landscape' is not empty, has 127 photos"
photoprism_1  | 2023-07-09T17:54:41.568424814Z time="2023-07-09T17:54:41Z" level=info msg="moments: label album 'Parks & Gardens' is not empty, has 422 photos"
photoprism_1  | 2023-07-09T17:54:42.023989979Z time="2023-07-09T17:54:42Z" level=debug msg="index: updating counts"
photoprism_1  | 2023-07-09T17:54:42.486533520Z time="2023-07-09T17:54:42Z" level=debug msg="counts: updated 2 places [462.504325ms]"
photoprism_1  | 2023-07-09T17:54:43.571567352Z time="2023-07-09T17:54:43Z" level=debug msg="counts: updated 4 subjects [1.084856282s]"
photoprism_1  | 2023-07-09T17:54:45.232296462Z time="2023-07-09T17:54:45Z" level=debug msg="counts: updated 1 label [1.660151603s]"
photoprism_1  | 2023-07-09T17:54:45.232387753Z time="2023-07-09T17:54:45Z" level=debug msg="index: updating covers"
photoprism_1  | 2023-07-09T17:54:45.644619028Z time="2023-07-09T17:54:45Z" level=debug msg="covers: updated 3 albums [412.213804ms]"
photoprism_1  | 2023-07-09T17:54:46.096513764Z time="2023-07-09T17:54:46Z" level=debug msg="covers: updated 0 folders [451.75548ms]"
photoprism_1  | 2023-07-09T17:54:46.297393533Z time="2023-07-09T17:54:46Z" level=debug msg="covers: updated 0 moments [201.036631ms]"
photoprism_1  | 2023-07-09T17:54:46.464365606Z time="2023-07-09T17:54:46Z" level=debug msg="covers: updated 0 months [166.758941ms]"
photoprism_1  | 2023-07-09T17:54:46.799324945Z time="2023-07-09T17:54:46Z" level=debug msg="covers: updated 0 states [334.760305ms]"
photoprism_1  | 2023-07-09T17:54:47.113897535Z time="2023-07-09T17:54:47Z" level=debug msg="covers: updated 0 countries [314.549251ms]"
photoprism_1  | 2023-07-09T17:54:47.119824350Z time="2023-07-09T17:54:47Z" level=debug msg="covers: updated 0 label moments [5.786777ms]"
photoprism_1  | 2023-07-09T17:54:47.145227423Z time="2023-07-09T17:54:47Z" level=debug msg="search: label plant includes 1 categories"
photoprism_1  | 2023-07-09T17:54:47.169258586Z time="2023-07-09T17:54:47Z" level=debug msg="photos: found 1 result for q:plants s:ar9vicm3l668ar05 filter:Plants public:true dist:20 count:1 order:newest [37.418878ms]"
photoprism_1  | 2023-07-09T17:54:47.351018740Z time="2023-07-09T17:54:47Z" level=debug msg="search: label animal includes 26 categories"
photoprism_1  | 2023-07-09T17:54:47.410023493Z time="2023-07-09T17:54:47Z" level=debug msg="photos: found 1 result for q:animal s:ar9wwanb05piv6bc filter:Animal public:true dist:20 count:1 order:newest [74.29414ms]"
photoprism_1  | 2023-07-09T17:54:47.601636045Z time="2023-07-09T17:54:47Z" level=debug msg="search: label flower includes 0 categories"
photoprism_1  | 2023-07-09T17:54:47.638480757Z time="2023-07-09T17:54:47Z" level=debug msg="photos: found 1 result for q:flowers s:ara4tdazofma2t3r filter:Flowers public:true dist:20 count:1 order:newest [47.597794ms]"
photoprism_1  | 2023-07-09T17:54:47.836353614Z time="2023-07-09T17:54:47Z" level=debug msg="covers: updated 3 smart albums [716.374898ms]"
photoprism_1  | 2023-07-09T17:54:47.836445085Z time="2023-07-09T17:54:47Z" level=warning msg="index: invalid ID specified (update covers)"
```

kvalev commented:
Lets move the cover-related discussion to the other PR.

As for the latest logs you sent me, only 191 out of 397 faces were matched with the latest approach (by looking for angled faces). I am not sure what the problem with the other images might be. If you want, you can try out adding a couple more angles to the config and see whether this helps:

https://github.com/kvalev/photoprism/pull/135/files#diff-9d687d9e5abd223347ebb1538e986d34bee6dbe22f7f25075ec6dea1e6bc0a3bR376

You can also (largely) debug the face detection locally using the `pigo` CLI. To do so, download the `facefinder` cascade file from [here](https://github.com/esimov/pigo/blob/master/cascade/facefinder) and then you can run the `pigo` face detection, which is used by PhotoPrism like that:
```
pigo -cf facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.7
```

The `scale`, `shift` and `iou` parameters you can set to the values above, as this is what PhotoPrism uses, while varying the `angle` to emulate different rotations. The `out.jpg` output image will contain bounding boxes with the detected faces. This way you can figure out additional rotation angles to hardcode.

Alternatively you can send me a new batch of photos to debug locally, especially if you find some photos where the faces were normally oriented, but still were not detected.

As for this PR, let me know whether you want to try and figure out couple more rotation angles to hardcode or if it is fine as-is and can be merged. (The additional angles can also be done in a separate PR and might even be changed to be configurable by the user and not just hardcoded, but his really depends on how many values can reasonably cover a large collection of face regions, such as yours).

rickysarraf commented:
> Lets move the cover-related discussion to the other PR.
> 

Sure. Thanks. 


> As for the latest logs you sent me, only 191 out of 397 faces were matched with the latest approach (by looking for angled faces). I am not sure what the problem with the other images might be. If you want, you can try out adding a couple more angles to the config and see whether this helps:
> 
> https://github.com/kvalev/photoprism/pull/135/files#diff-9d687d9e5abd223347ebb1538e986d34bee6dbe22f7f25075ec6dea1e6bc0a3bR376
> 
> You can also (largely) debug the face detection locally using the `pigo` CLI. To do so, download the `facefinder` cascade file from [here](https://github.com/esimov/pigo/blob/master/cascade/facefinder) and then you can run the `pigo` face detection, which is used by PhotoPrism like that:
> 
> ```
> pigo -cf facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.7
> ```
> 
> The `scale`, `shift` and `iou` parameters you can set to the values above, as this is what PhotoPrism uses, while varying the `angle` to emulate different rotations. The `out.jpg` output image will contain bounding boxes with the detected faces. This way you can figure out additional rotation angles to hardcode.
>

Thank you. I tried the example image with `pigo`.


The picture has 4 person/faces in total. Surprisingly, the default angle value for pigo resulted in 11 faces being reported. Of them 2 faces have 2 extra embeddings, while the remaining all are wrongly detected; as in they are not faces at all.


```
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.1
Detecting faces... ✔
6 face(s) detected

Execution time: 3.00s
12:25 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.2
Detecting faces... ✔
6 face(s) detected

Execution time: 3.04s
12:26 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.3
Detecting faces... ✔
2 face(s) detected

Execution time: 3.06s
12:26 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.4
Detecting faces... ✔
4 face(s) detected

Execution time: 2.92s
12:27 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.5
Detecting faces... ✔
8 face(s) detected

Execution time: 2.96s
12:27 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.6
Detecting faces... ✔
7 face(s) detected

Execution time: 3.09s
12:27 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.7
Detecting faces... ✔
11 face(s) detected

Execution time: 3.04s
12:27 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.8
Detecting faces... ✔
5 face(s) detected

Execution time: 3.04s
12:28 ♒ ॐ ♅ ♄ ⛢     ☺ 😄    
rrs@xps:~/VARTMP$ pigo -cf ~/go-facefinder -in 20230323_040840_21967141.jpg -out out.jpg -scale 1.1 -shift 0.1 -iou 0.4 -angle 0.9
Detecting faces... ✔
4 face(s) detected

Execution time: 2.92s
```

So, in neither of the invocations, was `pigo` able to correctly detect the faces.
 
> Alternatively you can send me a new batch of photos to debug locally, especially if you find some photos where the faces were normally oriented, but still were not detected.
>

This is fun. I'll use the same example runs on other images to make some sense. Whether it is really the picture where the faces are hard to detect, or if `pigo`  may need more refinements.

> As for this PR, let me know whether you want to try and figure out couple more rotation angles to hardcode or if it is fine as-is and can be merged. (The additional angles can also be done in a separate PR and might even be changed to be configurable by the user and not just hardcoded, but his really depends on how many values can reasonably cover a large collection of face regions, such as yours).

In the above example `pigo` run, none of the angels seems to help detect those faces.

Yes, Please. Lets merge it. But I'd humbly request making these knobs a tunable. Allows more flexibility without doing an entire rebuild. Infact, I use the YOLO threshold the same way, via a tunable.

```
      # Plugin for Yolo and realesrgan
      PHOTOPRISM_PLUGIN_YOLO8_ENABLED: "true"
      PHOTOPRISM_PLUGIN_YOLO8_HOSTNAME: "yolo8"
      PHOTOPRISM_PLUGIN_YOLO8_THRESHOLDCONFIDENCE: "0.3"
      PHOTOPRISM_PLUGIN_YOLO8_PORT: "5000"
      PHOTOPRISM_PLUGIN_REALESRGAN_ENABLED: "true"
      PHOTOPRISM_PLUGIN_REALESRGAN_HOSTNAME: "realesrgan"
      PHOTOPRISM_PLUGIN_REALESRGAN_PORT: "5001"
```

kvalev committed:
People: Configurable rotation angles

The user can now configure the rotation angles to be used for the fallback face
recognition for exif face regions, instead of using hardcoded values.
The environment variable is called PHOTOPRISM_FACE_REGION_ANGLES and takes in a
list of comma-separated rotation angles in degrees (0-360).

kvalev commented:
> Yes, Please. Lets merge it. But I'd humbly request making these knobs a tunable. Allows more flexibility without doing an entire rebuild.

Done. There are no hardcoded angles anymore, instead a list of angles can be configured using the `PHOTOPRISM_FACE_REGION_ANGLES` environment variable. However be aware that due to some limitations in the CLI framework used by PhotoPrism you have to configure the angles as integers in degrees and **not** as floats in the 0-1 range, as required by pigo. Converting between the two is trivial - just divide/multiply by 360, for example 90° corresponds to 0.25 in pigo, 0.7 in pigo equals to 252°.

rickysarraf commented:
> > Yes, Please. Lets merge it. But I'd humbly request making these knobs a tunable. Allows more flexibility without doing an entire rebuild.
> 
> Done. There are no hardcoded angles anymore, instead a list of angles can be configured using the `PHOTOPRISM_FACE_REGION_ANGLES` environment variable. However be aware that due to some limitations in the CLI framework used by PhotoPrism you have to configure the angles as integers in degrees and **not** as floats in the 0-1 range, as required by pigo. Converting between the two is trivial - just divide/multiply by 360, for example 90° corresponds to 0.25 in pigo, 0.7 in pigo equals to 252°.

Thank you for the quick changes @kvalev 

The tunables are being used correct and everything is working fine.

I set the different scan angles as:

```
      # Face Rotation Angles
      PHOTOPRISM_FACE_REGION_ANGLES: "03,36,72,108,144,180,216,252,288,324"
```

And the resultant log:

```
photoprism_1  | 2023-07-13T08:58:09.122189002Z time="2023-07-13T08:58:09Z" level=debug msg="media: opened 2023/07/20230709_105235_23D421A3.jpg [23 MB]"
photoprism_1  | 2023-07-13T08:58:09.281347814Z time="2023-07-13T08:58:09Z" level=info msg="media: created 8 thumbnails for 2023/07/20230709_105235_23D421A3.jpg [341.668525ms]"
photoprism_1  | 2023-07-13T08:58:09.289715236Z time="2023-07-13T08:58:09Z" level=debug msg="media: created no new thumbnails for 2023/07/20230709_105235_23D421A3.jpg [151.311µs]"
photoprism_1  | 2023-07-13T08:58:09.300989378Z time="2023-07-13T08:58:09Z" level=debug msg="index: found face region metadata in 2023/07/20230709_105235_23D421A3.jpg ([anshika-raj-sarraf-1923-12
38-261 matsya-raj-sarraf-2430-1161-261 priya-raj-sarraf-1376-345-554])"
photoprism_1  | 2023-07-13T08:58:09.301025636Z time="2023-07-13T08:58:09Z" level=debug msg="index: processing face region anshika-raj-sarraf-1923-1238-261"
photoprism_1  | 2023-07-13T08:58:09.475125961Z time="2023-07-13T08:58:09Z" level=debug msg="markers: cannot find face when uid is empty"
photoprism_1  | 2023-07-13T08:58:09.475170627Z time="2023-07-13T08:58:09Z" level=debug msg="index: successfully added face anshika-raj-sarraf-1923-1238-261 to file ''"
photoprism_1  | 2023-07-13T08:58:09.475180875Z time="2023-07-13T08:58:09Z" level=debug msg="index: processing face region matsya-raj-sarraf-2430-1161-261"
photoprism_1  | 2023-07-13T08:58:09.644588995Z time="2023-07-13T08:58:09Z" level=warning msg="index: no embeddings for face region matsya-raj-sarraf-2430-1161-261 and file 2023/07/20230709_10523
5_23D421A3.jpg, will try to recover"
photoprism_1  | 2023-07-13T08:58:09.644637516Z time="2023-07-13T08:58:09Z" level=debug msg="index: running face detection at angle 0.01"
photoprism_1  | 2023-07-13T08:58:14.930453626Z time="2023-07-13T08:58:14Z" level=debug msg="index: matched face face-2403-1383-605 has 0 embeddings"
photoprism_1  | 2023-07-13T08:58:15.102152105Z time="2023-07-13T08:58:15Z" level=debug msg="index: rotated face matsya-raj-sarraf-2430-1161-261 has 0 embeddings"
photoprism_1  | 2023-07-13T08:58:15.102187799Z time="2023-07-13T08:58:15Z" level=debug msg="index: running face detection at angle 0.10"
photoprism_1  | 2023-07-13T08:58:20.235377454Z time="2023-07-13T08:58:20Z" level=debug msg="index: matched face face-2441-1019-313 has 1 embedding"
photoprism_1  | 2023-07-13T08:58:20.236478029Z time="2023-07-13T08:58:20Z" level=debug msg="markers: cannot find face when uid is empty"
photoprism_1  | 2023-07-13T08:58:20.236498973Z time="2023-07-13T08:58:20Z" level=debug msg="index: successfully added face matsya-raj-sarraf-2430-1161-261 to file ''"
photoprism_1  | 2023-07-13T08:58:20.236530587Z time="2023-07-13T08:58:20Z" level=debug msg="index: processing face region priya-raj-sarraf-1376-345-554"
photoprism_1  | 2023-07-13T08:58:20.407447781Z time="2023-07-13T08:58:20Z" level=debug msg="markers: cannot find face when uid is empty"
photoprism_1  | 2023-07-13T08:58:20.407485250Z time="2023-07-13T08:58:20Z" level=debug msg="index: successfully added face priya-raj-sarraf-1376-345-554 to file ''"
photoprism_1  | 2023-07-13T08:58:20.638078774Z time="2023-07-13T08:58:20Z" level=debug msg="crop: 550eb379b146702b0fbfcd49be0b3feb05a1220e_720x720_fit.jpg is too small, upscaling 85px to 160px"
photoprism_1  | 2023-07-13T08:58:20.640389791Z time="2023-07-13T08:58:20Z" level=debug msg="crop: saved 550eb379b146702b0fbfcd49be0b3feb05a1220e_160x160_crop_25c2190d1076.jpg"
photoprism_1  | 2023-07-13T08:58:20.706813491Z time="2023-07-13T08:58:20Z" level=info msg="index: found 1 face in 20230709_105235_23D421A3.jpg [298.974851ms]"
photoprism_1  | 2023-07-13T08:58:21.124749072Z time="2023-07-13T08:58:21Z" level=info msg="index: matched 1 label with 20230709_105235_23D421A3.jpg [413.514853ms]"
photoprism_1  | 2023-07-13T08:58:25.337641438Z time="2023-07-13T08:58:25Z" level=debug msg="server: GET /api/v1/config (200) [162.231455ms]"
```
